shadows/framework/src/main/java/org/robolectric/shadows/ShadowParcelFileDescriptor.java

```diff
@@ -46,15 +49,32 @@ protected void __constructor__(ParcelFileDescriptor wrapped) {
   }
 
   @Implementation
-  protected static ParcelFileDescriptor open(File file, int mode) throws FileNotFoundException {
-    ParcelFileDescriptor pfd;
+  protected static ParcelFileDescriptor adoptFd(int fd) {
     try {
-      Constructor<ParcelFileDescriptor> constructor =
-          ParcelFileDescriptor.class.getDeclaredConstructor(FileDescriptor.class);
-      pfd = constructor.newInstance(new FileDescriptor());
+      File file = new File(Files.readSymbolicLink(Paths.get("/proc/self/fd/" + fd)).toString());
+      // TODO: how to support access mode create, truncate, append?
+      // It seems that we could not get those info from Java API.
+      // Should we keep a Map<int, RandomAccessFile> to map fd id to file?
+      boolean readable = file.canRead();
+      boolean writable = file.canWrite();
+      int mode = 0;
+      if (writable) {
+        mode = ParcelFileDescriptor.MODE_READ_WRITE;
+      } else if (readable) {
+        mode = ParcelFileDescriptor.MODE_READ_ONLY;
+      } else {
+        throw new IllegalStateException("No read and write permission");
+      }
+      return ParcelFileDescriptor.open(file, mode);
```

Comment:
I have some questions about shadowing adoptFd.

1. Is the original access mode associated with the fd needed? It seems it could not be supported when using pure Java API, while AOSP does it using native calls.
2. Is detachFd the only way to get the fd id for the adopt step? If so, I think maybe keeping a stateful mapping between fd id to  RandomAccessFile is a good solution?

@hoisie @utzcoz Any suggestions for this? Appreciate if you could help!

Reply:
> Is the original access mode associated with the fd needed?

It's better if we can support it. But if it is difficult to implement it with pure Java, IMO, we can leave it for discussion, and provide basic supporting at this PR if Google's internal test suites don't break with this PR. I think @hoisie can provide more information about it.

> Is detachFd the only way to get the fd id for the adopt step?

Looks like `detachFd` is used to transfer ownership of native fd, and https://cs.android.com/android/platform/superproject/+/master:frameworks/base/graphics/java/android/graphics/pdf/PdfEditor.java;l=83;bpv=1;bpt=1 also uses `getFd` to use fd.

Reply:
OK, I think portable PFD shadows implementation should be provided rather than OS specific implement(There are too much and too complex in native calls). Thus, probably some constraints on the fd that can be adopted are needed to simplify it,  for example only the fd from detachFd can be passed to adoptFd.